src/Functions/countMatches.h

```diff
@@ -83,13 +80,22 @@ class FunctionCountMatches : public IFunction
                 vec_res[i] = countMatches(str, re, matches);
             }
 
-            return result_column;
+            return col_res;
         }
-        else if (const ColumnConst * col_const_str = checkAndGetColumnConstStringOrFixedString(column_haystack))
+        else if (const ColumnFixedString * col_haystack_fixedstring = checkAndGetColumn<ColumnFixedString>(col_haystack))
         {
-            std::string_view str = col_const_str->getDataColumn().getDataAt(0).toView();
-            uint64_t matches_count = countMatches(str, re, matches);
-            return result_type->createColumnConst(input_rows_count, matches_count);
+            auto col_res = ColumnUInt64::create();
+
+            ColumnUInt64::Container & vec_res = col_res->getData();
+            vec_res.resize(input_rows_count);
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                std::string_view str = col_haystack_fixedstring->getDataAt(i).toView();
+                vec_res[i] = countMatches(str, re, matches);
+            }
+
+            return col_res;
         }
         else
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Error in FunctionCountMatches::getReturnTypeImpl()");
```

Comment:
This error message is not informative, it should indicate what went wrong, e.g. "Failed to cast haystack input to a string type (String, FixedString, a string literal)".

Reply:
We should theoretically not even end up here (`LOGICAL_ERRROR`). But you are right, I improved the wording.